core/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java

```diff
@@ -215,4 +337,109 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+        List<Object> terms;
+        if (isTermsLookupQuery()) {
+            terms = context.indexQueryParserService().handleTermsLookup(termsLookup);
+        } else {
+            terms = values;
+        }
+        if (terms == null || terms.isEmpty()) {
+            return Queries.newMatchNoDocsQuery();
+        }
+        return handleTermsQuery(terms, fieldName, context, minimumShouldMatch, disableCoord);
+    }
+
+    private static Query handleTermsQuery(List<Object> terms, String fieldName, QueryShardContext context, String minimumShouldMatch, boolean disableCoord) {
+        MappedFieldType fieldType = context.fieldMapper(fieldName);
+        String indexFieldName;
+        if (fieldType != null) {
+            indexFieldName = fieldType.names().indexName();
+        } else {
+            indexFieldName = fieldName;
+        }
+
+        Query query;
+        if (context.isFilter()) {
+            if (fieldType != null) {
+                query = fieldType.termsQuery(terms, context);
+            } else {
+                BytesRef[] filterValues = new BytesRef[terms.size()];
+                for (int i = 0; i < filterValues.length; i++) {
+                    filterValues[i] = BytesRefs.toBytesRef(terms.get(i));
+                }
+                query = new TermsQuery(indexFieldName, filterValues);
+            }
+        } else {
+            BooleanQuery bq = new BooleanQuery(disableCoord);
+            for (Object term : terms) {
+                if (fieldType != null) {
+                    bq.add(fieldType.termQuery(term, context), BooleanClause.Occur.SHOULD);
+                } else {
+                    bq.add(new TermQuery(new Term(indexFieldName, BytesRefs.toBytesRef(term))), BooleanClause.Occur.SHOULD);
+                }
+            }
+            Queries.applyMinimumShouldMatch(bq, minimumShouldMatch);
+            query = bq;
+        }
+        return query;
+    }
+
+    @Override
+    public QueryValidationException validate() {
+        QueryValidationException validationException = null;
+        if (this.fieldName == null) {
+            validationException = addValidationError("field name cannot be null.", validationException);
+        }
+        if (isTermsLookupQuery() && this.values != null) {
+            validationException = addValidationError("can't have both a terms query and a lookup query.", validationException);
+        }
+        if (isTermsLookupQuery()) {
+            QueryValidationException exception = termsLookup.validate();
+            if (exception != null) {
+                validationException = QueryValidationException.addValidationErrors(exception.validationErrors(), validationException);
+            }
+        }
+        return validationException;
+    }
+
+    @Override
+    protected TermsQueryBuilder doReadFrom(StreamInput in) throws IOException {
+        TermsQueryBuilder termsQueryBuilder = new TermsQueryBuilder(in.readString());
+        if (in.readBoolean()) {
```

Comment:
I understood from the last review that we were going to serialize everything all the time rather than using this conditional. Isn't that what you meant @cbuescher ?